### PR TITLE
test(amf): Additional Unit Test Cases for Stateless Feature

### LINKED ISSUE #10960: Unit Test for stateless feature

### PR DESCRIPTION
**Pull request by Krishnavamsi-wavelabs**
## Summary
**Added Unit Test Case for Stateless Feature**

## Test Plan
- Verified basic sanity with ueransim.
- Verified with Unit test.
- Please find attached logs and pcap for basic registration

MME LOG:
[mme.log](https://github.com/magma/magma/files/7793846/mme.log)

PCAP:
pcap

UNIT TEST:
test_oai

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by Krishnavamsi-wavelabs** [comment]
**PR Rebased:**

mme log: [05/01/2022]
[mme.log](https://github.com/magma/magma/files/7813722/mme.log)

Pcap: [05/01/2022]
pcap

Unit Test:
make_test_oai

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 1]
```diff
@@ -64,8 +64,7 @@ void amf_ue_context_update_coll_keys(
       ue_context_p->amf_context.imsi64,
       GUTI_ARG_M5G(&ue_context_p->amf_context._guti));
 
-  if ((gnb_ngap_id_key != INVALID_GNB_UE_NGAP_ID_KEY) &&
-      (ue_context_p->gnb_ngap_id_key != gnb_ngap_id_key)) {
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Why is this check removed?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 2]
```diff
@@ -84,20 +83,18 @@ void amf_ue_context_update_coll_keys(
   }
 
   if (amf_ue_ngap_id != INVALID_AMF_UE_NGAP_ID) {
-    if (ue_context_p->amf_ue_ngap_id != amf_ue_ngap_id) {
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:87]: Why is this check removed?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 3]
```diff
@@ -436,5 +499,323 @@ void AmfNasStateConverter::proto_to_amf_security_context(
...
+  state_smf_proc_data->max_downlink = smf_proc_data_proto.max_downlink();
+  state_smf_proc_data->pdu_session_type =
+      static_cast<M5GPduSessionType>(smf_proc_data_proto.pdu_session_type());
+  state_smf_proc_data->ssc_mode = smf_proc_data_proto.ssc_mode();
+}
+
+void AmfNasStateConverter::s_nssai_to_proto(
+    const s_nssai_t* state_s_nssai, magma::lte::oai::SNssai* snassi_proto) {
+  snassi_proto->set_sst(state_s_nssai->sst);
+  snassi_proto->set_sd(*(uint32_t*) &state_s_nssai->sd);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Please rebase this once #10958 is merged, since this code is changed there.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 4]
```diff
@@ -436,5 +499,323 @@ void AmfNasStateConverter::proto_to_amf_security_context(
...
+}
+void AmfNasStateConverter::proto_to_s_nssai(
+    const magma::lte::oai::SNssai& snassi_proto, s_nssai_t* state_s_nssai) {
+  state_s_nssai->sst              = snassi_proto.sst();
+  *(uint32_t*) &state_s_nssai->sd = snassi_proto.sd();
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Rebase on #10958 , which changes this code

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 5]
```diff
@@ -61,14 +61,20 @@ void clear_amf_nas_state() {
...
   return AmfNasStateManager::getInstance().get_ue_state_map();
 }
 
 void delete_amf_ue_state(imsi64_t imsi64) {
-  auto imsi_str =
-      magma5g::AmfNasStateManager::getInstance().get_imsi_str(imsi64);
-  magma5g::AmfNasStateManager::getInstance().clear_ue_state_db(imsi_str);
+  OAILOG_DEBUG(LOG_MME_APP, "Delete AMF ue state, %lu", imsi64);
+#if !MME_UNIT_TEST
+  auto imsi_str = AmfNasStateManager::getInstance().get_imsi_str(imsi64);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:72]: Please add comment here explaining that the removal is happening from the DB or from AMF client servicer

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 6]
```diff
@@ -436,5 +499,323 @@ void AmfNasStateConverter::proto_to_amf_security_context(
...
+      state_smf_context->gtp_tunnel_id.gnb_gtp_teid);
+
+  char gnb_gtp_teid_ip_addr_str[16] = {0};
+  inet_ntop(
+      AF_INET, state_smf_context->gtp_tunnel_id.gnb_gtp_teid_ip_addr,
+      gnb_gtp_teid_ip_addr_str, INET_ADDRSTRLEN);
+  smf_context_proto->set_gnb_gtp_teid_ip_addr(gnb_gtp_teid_ip_addr_str);
+
+  smf_context_proto->set_upf_gtp_teid(
+      *(uint32_t*) &state_smf_context->gtp_tunnel_id.upf_gtp_teid);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Rebase after #10958 is merged

**Review by ssanadhya — CHANGES_REQUESTED** [review]
First set of review comments.

**Comment by ssanadhya** [comment]
@Krishnavamsi-wavelabs , could you please rebase this PR on latest master, since #10958 is merged?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 1 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: - During initial Registratraton Message Processing, AMF is not able to update below hash tables
                       1. gnb_ue_ngap_id_ue_context_htbl
                       2. amf_state_ue_id_ht 
 
  Arguments of amf_ue_context_update_coll_keys api are passing UeContext Parameters  and Not-Equal check always fail.
 
**//amf_app_handler.cpp (line 170)**
     amf_ue_context_update_coll_keys(
        &amf_app_desc_p->amf_ue_contexts, ue_context_p,
        ue_context_p->gnb_ngap_id_key, ue_context_p->amf_ue_ngap_id,
        ue_context_p->amf_context.imsi64, ue_context_p->amf_teid_n11, guti_p);

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp (hunk 3 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Comment Addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp (hunk 6 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Comment Addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 5 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:72]: Comment Addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp (hunk 4 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Comment Addressed

**Comment by Krishnavamsi-wavelabs** [comment]
Thank You for reviewing @ssanadhya. 
I rebased PR with latest master.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 1 above)
Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Are ue_context_p->gnb_ngap_id_key and ue_context_p->amf_ue_ngap_id getting updated before the call to amf_ue_context_update_coll_keys? Ideally, during Initial Registration Message, these values should be invalid (passing the not equal check) and then get updated here.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp [hunk 7]
```diff
@@ -31,10 +31,9 @@ extern "C" {
 
 namespace magma5g {
 extern task_zmq_ctx_t amf_app_task_zmq_ctx;
+extern std::unordered_map<imsi64_t, guti_and_amf_id_t> amf_supi_guti_map;
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:35]: we should limit the usage of `extern` declarations, especially for mutable types such as maps, if so please add a comment where is this initialized, as I see this pattern is also being used on `amf_recv.cpp` 

Ideally these all should be retrieved either by passed as arguments or through a `get` public function, cc @ssanadhya

**Review by ardzoht — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 8]
```diff
@@ -105,6 +114,8 @@ int AmfNasStateManager::initialize_state(const amf_config_t* amf_config_p) {
 // TODO: This is a temporary check and will be removed in upcoming PR
 #if MME_UNIT_TEST
   read_state_from_db();
+  read_ue_state_from_db();
+  amf_sync_app_maps_from_db();
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:119]: Why do we need a sync call after `read_state_from_db` and `read_ue_state_from_db`?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 9]
```diff
@@ -156,18 +168,145 @@ amf_app_desc_t* AmfNasStateManager::get_state(bool read_from_redis) {
   state_dirty = true;
   if (persist_state_enabled && read_from_redis) {
     read_state_from_db();
+    read_ue_state_from_db();
+    amf_sync_app_maps_from_db();
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:174]: Why is this needed here?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 10]
```diff
@@ -156,18 +168,145 @@ amf_app_desc_t* AmfNasStateManager::get_state(bool read_from_redis) {
...
   return state_cache_p;
 }
 
-map_uint64_ue_context_t AmfNasStateManager::get_ue_state_map() {
-  return state_ue_map;
+map_uint64_ue_context_t* AmfNasStateManager::get_ue_state_map() {
+  return &state_ue_map;
+}
+
+/* Get the ue id from IMSI */
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Why is this part of state manager? It should be part of the `amf_app_ue_context.cpp`

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 11]
```diff
@@ -156,18 +168,145 @@ amf_app_desc_t* AmfNasStateManager::get_state(bool read_from_redis) {
...
+  if (redis_client->clear_keys(keys_to_del) != RETURNok) {
+    OAILOG_ERROR(LOG_AMF_APP, "Failed to clear the state in data store");
+    return;
+  }
+}
+
+void put_amf_ue_state(
+    amf_app_desc_t* amf_app_desc_p, imsi64_t imsi64, bool force_ue_write) {
+  if (AmfNasStateManager::getInstance().is_persist_state_enabled()) {
+    if (imsi64 != INVALID_IMSI64) {
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: For ease of readability

```suggestion
 if ((!AmfNasStateManager::getInstance().is_persist_state_enabled()) || (imsi64 == INVALID_IMSI64)) {
    return
  }
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 12]
```diff
@@ -156,18 +168,145 @@ amf_app_desc_t* AmfNasStateManager::get_state(bool read_from_redis) {
...
+    OAILOG_DEBUG(
+        log_task, "Finished writing UE state for IMSI %s", imsi_str.c_str());
+  }
+#endif
+}
+
+status_code_e AmfNasStateManager::read_ue_state_from_db() {
+#if !MME_UNIT_TEST
+  /* Data store is Redis db. In this case actual call is made to Redis db */
+  // StateManager::read_ue_state_from_db();
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Remove commented code

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 13]
```diff
@@ -156,18 +168,145 @@ amf_app_desc_t* AmfNasStateManager::get_state(bool read_from_redis) {
...
+  if (!persist_state_enabled) {
+    return RETURNok;
+  }
+  auto keys = redis_client->get_keys("IMSI*" + task_name + "*");
+  for (const auto& key : keys) {
+    magma::lte::oai::UeContext ue_proto = magma::lte::oai::UeContext();
+    ue_m5gmm_context_t* ue_context_p    = new ue_m5gmm_context_t();
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Move declaration closer to reference, so that no allocation is done if condition in line 270 fails.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp [hunk 14]
```diff
@@ -121,6 +125,210 @@ TEST(TestAMFStateConverter, TestStateToProto) {
...
+      ue_m5gmm_context1.paging_context.paging_retx_count,
+      ue_m5gmm_context2.paging_context.paging_retx_count);
+}
+
+TEST(TestAMFStateConverter, TestAmfContextStateToProto) {
+#define AMF_CAUSE_SUCCESS 1
+  amf_context_t amf_ctx1 = {}, amf_ctx2 = {};
+  magma::lte::oai::EmmContext emm_context_proto = magma::lte::oai::EmmContext();
+
+  amf_ctx1.imsi64             = 222456000000101;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp:208]: Why is this value different from the digits of amf_ctxt1.imsi.u in lines 209-223 ?

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp [hunk 15]
```diff
@@ -447,14 +655,39 @@ class AMFAppStatelessTest : public ::testing::Test {
...
   virtual void TearDown() {
     clear_amf_nas_state();
     clear_amf_config(&amf_config);
     destroy_task_context(&amf_app_task_zmq_ctx);
     itti_free_desc_threads();
+    AMFClientServicer::getInstance().map_table_key_proto_str.clear();
+    AMFClientServicer::getInstance().map_imsi_ue_proto_str.clear();
+  }
+
+  // This Function mocks mme restart.
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
  // This Function mocks AMF task restart.
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp [hunk 16]
```diff
@@ -447,14 +655,39 @@ class AMFAppStatelessTest : public ::testing::Test {
...
+    for_each(ue_context_map.begin(), ue_context_map.end(), [&](auto& it) {
+      ue_id_vector.push_back(it.first);
+    });
+    for (auto& ue_id : ue_id_vector) {
+      ue_m5gmm_context_t* ue_context_p =
+          amf_ue_context_exists_amf_ue_ngap_id((amf_ue_ngap_id_t) ue_id);
+      if (ue_context_p) {
+        // Clean up the procedures
+        amf_nas_proc_clean_up(ue_context_p);
+        delete ue_context_p;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Use [unordered_map::erase](https://www.cplusplus.com/reference/unordered_map/unordered_map/erase/) to delete `ue_context_p` as you iterate over the map.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp [hunk 17]
```diff
@@ -447,14 +655,39 @@ class AMFAppStatelessTest : public ::testing::Test {
...
+      }
+    }
+    ue_context_map.clear();
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Not needed if you use `erase`

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp (hunk 17 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Comment Addressed

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp (hunk 16 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Comment Addressed

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp (hunk 15 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Comment Addressed

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp (hunk 14 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp:208]: Comment Addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 13 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Comment Addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 12 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Comment Addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 11 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Comment Addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 10 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Comment Addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 8 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:119]: In the existing AMF implementation there are two sets of hashtables:
1. Defined in amf_app_state_manager. 
2. Defined in the application code (amf_supi_guti_map, ue_context_map).

So, this amf_sync_app_maps_from_db() api syncs the state_manager maps with the application maps.
This is a duplicate effort, which will be removed as part of PR (Fixing the global maps #11087).

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp (hunk 7 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:35]: This is a duplicate effort, these maps will be removed as part of PR (Fixing the global maps #11087).

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 1 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Yes, I agree with you. In the existing implementation it was getting updated before the call to amf_ue_context_update_coll_keys.
 Have made the required changes.
This surfaced a known issue: IMSI is not populated incase of SUCI Registration. 
So, the test case TestRegistrationProcSUCIExt is not passing. This issue is being tracked here #11382.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 9 above)
Comment (ardzoht) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:174]: +1

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp [hunk 18]
```diff
@@ -836,6 +836,7 @@ TEST_F(AMFAppProcedureTest, TestPDUSession_Invalid_PDUSession_Identity) {
   EXPECT_TRUE(expected_Ids == AMFClientServicer::getInstance().msgtype_stack);
 }
 
+#if 0
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:841]: what is the purpose of this?

Comment (ganeshg87) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:841]: there is a known issue,  TestRegistrationProcSUCIExt test case is failing, we have commented test case. 
This issue is being tracked here #11382, will be assigned to the module owner.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 9 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:174]: - AmfNasStateManager has <ue_id, ue_context> map(amf_app_state_manager.cpp) 
- and amf also has another global map  <ue_id, ue_context> map(amf_app_ue_context.cpp) 
#11087 PR addressed these duplicate maps issue.

stateless feature sync only AmfNasStateManager data structures, amf_sync_app_maps_from_db was added to sync amf ue_context_map defined in amf_app_ue_context.cpp. this will be removed after #11087 merge.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 8 above)
Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:119]: @Krishnavamsi-wavelabs , could you please add a `TODO` comment here with @panyogesh 's name here so that this is indeed removed as part of #11087

Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:119]: Comment Addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 19]
```diff
@@ -104,10 +103,9 @@ void amf_ue_context_update_coll_keys(
         amf_ue_ngap_id);
   }
 
-  m_rc = amf_ue_context_p->imsi_amf_ue_id_htbl.remove(
-      ue_context_p->amf_context.imsi64);
-
-  if (INVALID_AMF_UE_NGAP_ID != amf_ue_ngap_id) {
+  if (INVALID_AMF_UE_NGAP_ID != amf_ue_ngap_id && imsi != 0) {
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]:
```suggestion
  if ((amf_ue_ngap_id != INVALID_AMF_UE_NGAP_ID) && (imsi != 0)) {
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 20]
```diff
@@ -134,20 +132,25 @@ void amf_ue_context_update_coll_keys(
...
+         (guti_p->guamfi.plmn.mcc_digit2 !=
+          ue_context_p->amf_context.m5_guti.guamfi.plmn.mcc_digit2) ||
+         (guti_p->guamfi.plmn.mcc_digit3 !=
+          ue_context_p->amf_context.m5_guti.guamfi.plmn.mcc_digit3)) ||
+        (ue_context_p->amf_ue_ngap_id != INVALID_AMF_UE_NGAP_ID) &&
+            ((guti_p->guamfi.amf_set_id != 0) &&
+             (guti_p->guamfi.amf_regionid != 0) && (guti_p->m_tmsi != 0) &&
+             (guti_p->guamfi.plmn.mcc_digit1 != 0) &&
+             (guti_p->guamfi.plmn.mcc_digit2 != 0) &&
+             (guti_p->guamfi.plmn.mcc_digit3 != 0))) {
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: MCC can have 0 digit in it, e.g. test networks start with digit 0

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 21]
```diff
@@ -134,20 +132,25 @@ void amf_ue_context_update_coll_keys(
...
       if (INVALID_AMF_UE_NGAP_ID != amf_ue_ngap_id) {
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Not related with this PR, but please fix the style here too
```suggestion
      if (amf_ue_ngap_id != INVALID_AMF_UE_NGAP_ID ) {
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp [hunk 22]
```diff
@@ -462,4 +461,33 @@ void amf_free_ue_context(ue_m5gmm_context_s* ue_context_p) {
   ue_context_p = NULL;
 }
 
+// Sync State manager map with Amf Application maps
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:466]: Add the `TODO` comment here too for @panyogesh

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp (hunk 18 above)
Comment (ssanadhya) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:841]: @Krishnavamsi-wavelabs , please add a `TODO` comment here referring this PR #11034 as the starting point of failure and that issue https://github.com/magma/magma/issues/11382 is tracking this

**Review by ssanadhya — APPROVED** [review]
LGTM, with minor comments. 
Can be merged with @ardzoht 's approval.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp [hunk 23]
```diff
@@ -621,6 +689,30 @@ void AmfNasStateConverter::proto_to_qos_flow_setup_request_item(
       qos_flow_item_proto.qos_flow_param(),
       &state_qos_flow_request_item->qos_flow_level_qos_param);
 }
+
+void AmfNasStateConverter::smf_context_map_to_proto(
+    const std::unordered_map<uint8_t, std::shared_ptr<smf_context_t>>
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]:
```suggestion
    const std::unordered_map<uint8_t, std::shared_ptr<smf_context_t>>&
```
pass by const reference

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h [hunk 24]
```diff
@@ -142,6 +142,26 @@ class AmfNasStateConverter : public magma::lte::StateConverter {
       const EmmSecurityContext& emm_security_context_proto,
       amf_security_context_t* state_amf_security_context);
 
+  static void guti_m5_to_proto(
+      const guti_m5_t& state_guti_m5, magma::lte::oai::Guti_m5* guti_m5_proto);
+  static void proto_to_guti_m5(
+      const magma::lte::oai::Guti_m5& guti_m5_proto, guti_m5_t* state_guti_m5);
+
+  static void smf_context_map_to_proto(
+      const std::unordered_map<uint8_t, std::shared_ptr<smf_context_t>>
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: pass by const reference

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp (hunk 7 above)
Comment (ardzoht) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:35]: Please add a `TODO` comment to specify whis will be removed on #11087

**Review by ardzoht — APPROVED** [review]
lgtm, left minor comments, please address before merging

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 2 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:87]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp (hunk 7 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:35]: Addressed Review Comment

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp (hunk 18 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:841]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 19 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 21 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp (hunk 23 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h (hunk 24 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp (hunk 22 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:466]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 20 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Addressed Review Comment

**Comment by Krishnavamsi-wavelabs** [comment]
Thank you @ardzoht and @ssanadhya for reviewing.
Addressed all the review comments and marked as ready to merge.
